README.md
# OASIS [![CircleCI](https://circleci.com/gh/HermitSun/oasis.svg?style=svg)](https://circleci.com/gh/HermitSun/oasis)

[![Codacy Badge](https://app.codacy.com/project/badge/Grade/60d45ecbcf1f47cbb145447bc5bd89f6)](https://www.codacy.com/manual/HermitSun/oasis?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=HermitSun/oasis&amp;utm_campaign=Badge_Grade)

## 使用说明

``` bash
# 安装依赖
$ npm install

# 在localhost:3000启动开发服务器
$ npm run dev

# 打包并运行
$ npm run build
$ npm run start

# 生成静态站点
$ npm run generate

# 因为服务器性能过低，无法完成CI，故采用这种方式
# 发布到wensun.top
$ npm run release

# 重新构建element-ui主题
# 注意，因为依赖过旧，需要node版本 < 12
# 建议使用10.x LTS
$ npm run theme
```


CHANGELOG.md
从3.1版本起，对每个版本的变更进行跟踪。

---

# 4.4 (2020-08-15)

### Code Refactoring

- 移除gitlab相关配置

### Features

- 搜索支持回车进行过滤(#5)

# 4.3.3 (2020-08-14)

### Bug Fixes

- 修复学术关系图可能会展示非当前学者的学术关系，并且有时会无法回退的问题(#3)
- 修复部分机构画像页无法加载的问题(#4)

### Performance Improvements

- 调整学术关系图缓存的实现方式，避免了可能的内存泄漏问题

# 4.3.2 (2020-08-13)

### Bug Fixes

- 修复排名页展开用户详情时切换排序标准，已展开内容不刷新的问题(#1)
  - 迁移到GitHub后issue序号重新计算

### Code Refactoring

- 迁移到circle-ci

### Performance Improvements

- 调整首页服务端渲染和客户端渲染的比重，提高并发性
- 移除无效的中间件

# 4.3.1 (2020-07-30)

### Bug Fixes

- 修复人才库里没有机构的作者的问题
- 修复按钮中的中文问题

### Performance Improvements

- 移除无效的依赖和组件
- 减少推荐请求数量

### Features

- 添加排名详情的空数据提示

# 4.3 (2020-07-30)

### Bug Fixes

- 修复点击推荐机构时出现400的问题
- 修复在高级搜索页面出现filter的问题
- 修复searchBar遮挡高级搜索的问题
- 修复高级排名Time Range存在歧义的问题
- 修复首页轮播的箭头覆盖高级搜索框的问题(#75)

### Code Refactoring

- 完善移动端适配
- 调整学者关系图样式

### Features

- 添加移动端的menu

# 4.2.1 (2020-07-30)

### Bug Fixes

- 修复网络图返回上一层的行为出现错误的问题(#73)
- 修复切换搜索关键词时filter不更新的问题

### Code Refactoring

- 移动端适配
- 部分样式调整
- 将普通搜索的filter移动到drawer中

### Performance Improvements

- 移除无效的依赖和组件
- 减少推荐请求数量

### Features

- 增加普通搜索推荐

# 4.2.0 (2020-07-30)

### Bug Fixes

- 修复学者关系图回退失败的问题(#72)

### Performance Improvements

- 点击当前结点时不再触发关系图重绘

### Features

- 浏览器后退键可以返回关系图上一层

# 4.1.4 (2020-07-29)

### Bug Fixes

- 修复排名页高亮显示不正确的问题(#68)
- 修复页面重入时不显示图表的问题
- 修复关系图中权重过低学者在图上消失的问题

### Code Refactoring

- 增加到机构画像的跳转方式(#69)
- 增加近十年无数据的placeholder(#70)
- 完善类型定义
- 重构关系图的实现

### Performance Improvements

- 移除d3pie，删除一些无用或冗余的组件
- 移除图表的resize以减少重绘

# 4.1.3 (2020-07-28)

### Bug Fixes

- 修复画像页和排名页的滚动条溢出

### Performance Improvements

- 懒加载首页的轮播图
- 懒加载标签页

# hotfix/trend-cannot-display (2020-07-28)

### Bug Fixes

- 修复趋势图无法显示的问题

# 4.1.2 (2020-07-28)

### Features

- 完成首页

### Performance Improvements

- 压缩图片

# 4.1.1 (2020-07-27)

### Performance Improvements

- 巨幅减小打包体积

# 4.1 (2020-07-27)

### Bug Fixes

- 修复图标无法显示的问题
- 修复搜索不高亮的问题
- 修复bar的x轴数据问题
- 修复ranking排序标准的bug
- 修复因为npm run theme带来的样式问题
- 修复CDN导入echarts顺序错误的问题

### Code Refactoring

- 调整echarts导入方式为CDN

### Features

- 添加画像页论文计数
- 添加人才库目录

### BREAKING CHANGES

- 将用户端与管理端分离

# 4.0.1 (2020-07-24)

### Bug Fixes

- 修复nav中的路径错误

### Code Refactoring

- 调整样式

# 4.0 (2020-07-23)

### Code Refactoring

- 将部分图表切换为echarts

### BREAKING CHANGES

- 用户端样式完全重构

# 3.3 (2020-05-28)

### Bug Fixes

- 修复公共页面缺乏样式文件的问题
- 修复echarts加载有时会报错的问题(#65)

### Code Refactoring

- 优化人才库样式

### Features

- 完成人才库页面
- 增加3D trend

# 3.2.2 (2020-05-21)

### Bug Fixes

- 修复数据导入初始进度会变成NaN的问题(#63)
- 修复数据导入清空日期后URL里只有date的问题(#64)

### Code Refactoring

- 优化基础人才库样式
- 对首页进行重构

### Features

- 数据导入在每个页面都增加新建任务按钮(#60)
- 增加数据导入任务开始时的提示

### BREAKING CHANGES

- 首页的样式发生了重大变化

# 3.2.1 (2020-05-20)

### Bug Fixes

- 修复研究方向画像页点击关键词页面不刷新的问题(#58)

### Code Refactoring

- 对画像页进行重构
- 按字典序引入components
- 简化设置默认值的方式（三目运算符 -> `||`）

### Performance Improvements

- 画像页不再重复渲染输入框和加载词云（重构）

# 3.2 (2020-05-20)

### Bug Fixes

- 实现可能出现内存泄漏的问题
- 修复SearchSortKeyComp的emit方法错误

### Code Refactoring

- 对数据导入页进行重构

### Features

- 增加管理端数据导入功能
- 增加命令搜索功能

### Performance Improvements

- 提高SEO权重（`nofollow`）
- 排名页采用更轻量的缓存实现

# 3.1.1 (2020-05-16)

### Code Refactoring

- 对排名页进行重构

### Performance Improvements

- 对排名页进行缓存

# 3.1 (2020-05-16)

### Bug Fixes

- 请求取消后不再提示Error(#54)
- 首页弹出取消请求信息(#55)
- 排名页无法展开(#56)

### Code Refactoring

- 完善词云加载完成的判断条件
- 改用`Promise.all`进行并发请求
- 补充研究方向画像的类型

### Features

- 增加研究方向排名
- 增加echarts基本配置

### Performance Improvements

- 减少学者排名页图表的阻塞时间


api/index.ts
import axios from './config';
import {
  AdvancedSearchPayload,
  BasicSearchPayload,
  CommandSearchPayload,
  FilterSearchPayload
} from '~/interfaces/requests/search/SearchPayload';
import { BasicResponse } from '~/interfaces/responses/BasicResponse';
import {
  SearchFullResponse,
  SearchReference
} from '~/interfaces/responses/search/SearchResponse';
import { BasicRankingResponse } from '~/interfaces/responses/ranking/basic/BasicRankingResponse';
import {
  KeywordRankingAdvancedPayload,
  RankingAdvancedPayload,
  RankingBasicPayload
} from '~/interfaces/requests/ranking/RankingPayload';
import { AuthorBasicRankingResponse } from '~/interfaces/responses/ranking/basic/AuthorBasicRankingResponse';
import { ResearcherInterestPayload } from '~/interfaces/requests/interest/ResearcherInterestPayload';
import { InterestResponse } from '~/interfaces/responses/interest/InterestResponse';
import { ActivePaperAbstractResponse } from '~/interfaces/responses/abstract/ActivePaperAbstractResponse';
import { SearchFilterPayload } from '~/interfaces/requests/search/SearchFilterPayload';
import { SearchFilterResponse } from '~/interfaces/responses/search/SearchFilterResponse';
import { PortraitResponse } from '~/interfaces/responses/portrait/PortraitResponse';
import { AffiliationPapersPayload } from '~/interfaces/requests/portrait/affiliation/AffiliationPaperPayload';
import {
  AuthorAdvancedRankingResponse,
  AuthorDetailRankingResponse
} from '~/interfaces/responses/ranking/advanced/AuthorAdvancedRankingResponse';
import {
  AffiliationAdvancedRankingResponse,
  AffiliationDetailRankingResponse
} from '~/interfaces/responses/ranking/advanced/AffiliationAdvancedRankingResponse';
import {
  AcademicRelationResponse,
  AuthorPortraitResponse
} from '~/interfaces/responses/portrait/AuthorResponse';
import { AuthorPapersPayload } from '~/interfaces/requests/portrait/author/AuthorPaperPayload';
import { KeywordPapersPayload } from '~/interfaces/requests/portrait/keyword/KeywordPaperPayload';
import {
  KeywordAdvancedRankingResponse,
  KeywordDetailRankingResponse
} from '~/interfaces/responses/ranking/advanced/KeywordAdvancedRankingResponse';
import { ActiveTalentsBaseResponse } from '~/interfaces/responses/talent/ActiveTalentsBaseResponse';
import { TalentsListResponse } from '~/interfaces/responses/talent/TalentsListResponse';
import { Keyword3DTrendResponse } from '~/interfaces/responses/charts/3DTrendResponse';
import { SearchRecommendAuthorsResponse } from '~/interfaces/responses/search/SearchRecommendAuthorsResponse';
import { SearchRecommendAffiliationsResponse } from '~/interfaces/responses/search/SearchRecommendAffiliationsResponse';

// 1. 普通搜索
export async function basicSearch(
  args: BasicSearchPayload
): Promise<BasicResponse<SearchFullResponse>> {
  const { data } = await axios.get('/search/basic/es/highlight', {
    params: args
  });
  return data;
}

// 1.1 二次搜索（模式：关键词分词，其他精确匹配）
export async function basicFilterSearch(
  args: FilterSearchPayload
): Promise<BasicResponse<SearchFullResponse>> {
  const { data } = await axios.get('/search/filter/es', {
    params: args
  });
  return data;
}

// 2. 组合搜索
export async function advancedSearch(
  args: AdvancedSearchPayload
): Promise<BasicResponse<SearchFullResponse>> {
  const { data } = await axios.get('/search/advanced/es/highlight', {
    params: args
  });
  return data;
}

// 3.1 查看机构论文简略排名
export async function getAffiliationBasicRanking(
  args: RankingBasicPayload
): Promise<BasicResponse<BasicRankingResponse[]>> {
  const { data } = await axios.get('/rank/basic/affiliation', {
    params: args
  });
  return data;
}

// 3.2 查看学者论文简略排名
export async function getAuthorBasicRanking(
  args: RankingBasicPayload
): Promise<BasicResponse<AuthorBasicRankingResponse[]>> {
  const { data } = await axios.get('/rank/basic/author', {
    params: args
  });
  return data;
}

// 3.3 查看会议论文简略排名
export async function getConferenceBasicRanking(
  args: RankingBasicPayload
): Promise<BasicResponse<BasicRankingResponse[]>> {
  const { data } = await axios.get('/rank/basic/conference', {
    params: args
  });
  return data;
}

// 3.4 查看期刊论文简略排名
export async function getJournalBasicRanking(
  args: RankingBasicPayload
): Promise<BasicResponse<BasicRankingResponse[]>> {
  const { data } = await axios.get('/rank/basic/journal', {
    params: args
  });
  return data;
}

// 3.5 查看研究方向论文简略排名
export async function getKeywordBasicRanking(
  year: number
): Promise<BasicResponse<BasicRankingResponse[]>> {
  const { data } = await axios.get('/rank/basic/keyword', {
    params: { year }
  });
  return data;
}

// 4. 查看学者研究方向
export async function getResearcherInterest(
  args: ResearcherInterestPayload
): Promise<BasicResponse<InterestResponse[]>> {
  const { data } = await axios.get('/researcher/interest', {
    params: args
  });
  return data;
}

// 5. 查看活跃论文摘要
export async function getActivePaperAbstract(): Promise<
  BasicResponse<ActivePaperAbstractResponse[]>
> {
  const { data } = await axios.get('/paper/abstract');
  return data;
}

// 7. 根据论文id获得reference
export async function getReferenceById(
  paperId: string
): Promise<BasicResponse<SearchReference[]>> {
  const { data } = await axios.get('/paper/reference', {
    params: { paperId }
  });
  return data;
}

// 8. 获取普通搜索的二次筛选条件
export async function getBasicSearchFilterCondition(
  args: SearchFilterPayload
): Promise<BasicResponse<SearchFilterResponse>> {
  const { data } = await axios.get('/search/basic/mongo/filter', {
    params: args
  });
  return data;
}

// 9. 查看作者排名详情
export async function getAuthorAdvancedRanking(
  args: RankingAdvancedPayload
): Promise<BasicResponse<AuthorAdvancedRankingResponse[]>> {
  const { data } = await axios.get('/rank/advanced/author', {
    params: args
  });
  return data;
}

// 10. 根据作者id获取排名细节
export async function getAuthorDetailRankingById(
  authorId: string
): Promise<BasicResponse<AuthorDetailRankingResponse>> {
  const { data } = await axios.get('/rank/detail/author', {
    params: { authorId }
  });
  return data;
}

// 11. 查看机构排名详情
export async function getAffiliationAdvancedRanking(
  args: RankingAdvancedPayload
): Promise<BasicResponse<AffiliationAdvancedRankingResponse[]>> {
  const { data } = await axios.get('/rank/advanced/affiliation', {
    params: args
  });
  return data;
}

// 12. 根据机构id获取排名细节
export async function getAffiliationDetailRankingById(
  affiliationId: string
): Promise<BasicResponse<AffiliationDetailRankingResponse>> {
  const { data } = await axios.get('/rank/detail/affiliation', {
    params: { affiliationId }
  });
  return data;
}

// 13.  根据作者id获取作者画像
export async function getAuthorPortraitById(
  authorId: string
): Promise<BasicResponse<AuthorPortraitResponse>> {
  const { data } = await axios.get('/portrait/author', {
    params: { authorId }
  });
  return data;
}

// 14. 根据作者id获取该作者发表的所有论文
export async function getAuthorPapersById(
  args: AuthorPapersPayload
): Promise<BasicResponse<SearchFullResponse>> {
  const { data } = await axios.get('/paper/author', {
    params: args
  });
  return data;
}

// 15. 根据作者id获取该作者的学者关系
export async function getAcademicRelationByAuthorId(
  id: string
): Promise<BasicResponse<AcademicRelationResponse>> {
  const { data } = await axios.get('/picture/academic', {
    params: { id }
  });
  return data;
}

// 16. 获取某机构画像
export async function getAffiliationPortrait(
  affiliation: string
): Promise<BasicResponse<PortraitResponse>> {
  const { data } = await axios.get('/portrait/affiliation', {
    params: { affiliation }
  });
  return data;
}

// 17. 获取某机构发表的所有论文
export async function getAffiliationPapers(
  args: AffiliationPapersPayload
): Promise<BasicResponse<SearchFullResponse>> {
  const { data } = await axios.get('/paper/affiliation', {
    params: args
  });
  return data;
}

// 18. 查看某机构研究方向
export async function getAffiliationInterest(
  affiliation: string
): Promise<BasicResponse<InterestResponse[]>> {
  const { data } = await axios.get('/affiliation/interest', {
    params: { affiliation }
  });
  return data;
}

// 19. 查看某机构下的作者排名详情
export async function getAuthorDetailRanking(
  affiliation: string
): Promise<BasicResponse<AuthorAdvancedRankingResponse[]>> {
  const { data } = await axios.get('/rank/affiliation/author', {
    params: { affiliation }
  });
  return data;
}

// 20. 获取某研究方向画像
export async function getKeywordPortrait(
  keyword: string
): Promise<BasicResponse<PortraitResponse>> {
  const { data } = await axios.get('/portrait/keyword', {
    params: { keyword }
  });
  return data;
}

// 21. 获取某研究方向发表的所有论文
export async function getKeywordPapers(
  args: KeywordPapersPayload
): Promise<BasicResponse<SearchFullResponse>> {
  const { data } = await axios.get('/paper/keyword', {
    params: args
  });
  return data;
}

// 22. 获取某会议画像
export async function getConferencePortrait(
  conference: string
): Promise<BasicResponse<PortraitResponse>> {
  const { data } = await axios.get('/portrait/conference', {
    params: { conference }
  });
  return data;
}

//  23. 获取某会议研究方向
export async function getConferenceInterest(
  conference: string
): Promise<BasicResponse<InterestResponse[]>> {
  const { data } = await axios.get('/conference/interest', {
    params: { conference }
  });
  return data;
}

// 24. 获取某期刊画像
export async function getJournalPortrait(
  journal: string
): Promise<BasicResponse<PortraitResponse>> {
  const { data } = await axios.get('/portrait/journal', {
    params: { journal }
  });
  return data;
}

// 25. 获取某期刊研究方向
export async function getJournalInterest(
  journal: string
): Promise<BasicResponse<InterestResponse[]>> {
  const { data } = await axios.get('/journal/interest', {
    params: { journal }
  });
  return data;
}

// 26和33为管理端共用接口
// 26. 获取机构信息
export async function getAffiliationInfo(
  name: string,
  page = 1
): Promise<BasicResponse<SearchRecommendAffiliationsResponse>> {
  const { data } = await axios.get('/info/affiliation', {
    params: {
      name,
      page
    }
  });
  return data;
}
// 33. 获取作者信息
export async function getAuthorInfo(
  name: string,
  page = 1
): Promise<BasicResponse<SearchRecommendAuthorsResponse>> {
  const { data } = await axios.get('/info/author', {
    params: {
      name,
      page
    }
  });
  return data;
}

// 35. 命令搜索
export async function commandSearch(
  args: CommandSearchPayload
): Promise<BasicResponse<SearchFullResponse>> {
  const { data } = await axios.get('/search/command', {
    params: args
  });
  return data;
}

// 38. 查看研究方向排名详情
export async function getKeywordAdvancedRanking(
  args: RankingAdvancedPayload
): Promise<BasicResponse<KeywordAdvancedRankingResponse[]>> {
  const { data } = await axios.get('/rank/advanced/keyword', {
    params: args
  });
  return data;
}

// 39.获取某研究方向下的作者排名详情
export async function getAuthorDetailRankingByKeyword(
  args: KeywordRankingAdvancedPayload
): Promise<BasicResponse<AuthorAdvancedRankingResponse[]>> {
  const { data } = await axios.get('/rank/detail/author/keyword', {
    params: args
  });
  return data;
}

// 40. 获取某研究方向下的机构排名详情
export async function getAffiliationDetailRankingByKeyword(
  args: KeywordRankingAdvancedPayload
): Promise<BasicResponse<AffiliationAdvancedRankingResponse[]>> {
  const { data } = await axios.get('/rank/detail/affiliation/keyword', {
    params: args
  });
  return data;
}

// 42. 合并关键词信息 mergeKeywords
export async function mergeKeywords(
  src: string[],
  dest: string
): Promise<BasicResponse> {
  const { data } = await axios.put('/info/keywords', { src, dest });
  return data;
}

// 43. 查看活跃人才库
export async function getActiveTalentsBase(): Promise<
  BasicResponse<ActiveTalentsBaseResponse[]>
> {
  const { data } = await axios.get('/talents/active');
  return data;
}

// 44. 1 查看某人才库的人才列表
export async function getTalentsListByTalentBase(
  field: string,
  page: number
): Promise<BasicResponse<TalentsListResponse[]>> {
  const { data } = await axios.get('/talents/list', {
    params: { field, page }
  });
  return data;
}

// 44.2 查看某人才库的
export async function getTalentsActivePapersByTalentBase(
  field: string
): Promise<BasicResponse<ActivePaperAbstractResponse[]>> {
  const { data } = await axios.get('/talents/recommend', {
    params: { field }
  });
  return data;
}

// 45. 获取某学者的相关学者推荐
export async function getSimilarAuthorByAuthorId(
  authorId: string
): Promise<BasicResponse<TalentsListResponse[]>> {
  const { data } = await axios.get('/talents/list', {
    params: { authorId }
  });
  return data;
}

// 48. 根据研究方向获取排名详情
export async function getKeywordDetailRanking(
  keyword: string
): Promise<BasicResponse<KeywordDetailRankingResponse>> {
  const { data } = await axios.get('/rank/detail/keyword', {
    params: { keyword }
  });
  return data;
}

// 49. 查看研究方向的排名3D趋势
export async function getKeyword3DTrend(): Promise<
  BasicResponse<Keyword3DTrendResponse>
> {
  const { data } = await axios.get('/keyword/trend');
  return data;
}
